### PR TITLE
Issue #8801: Support for all Token Types as a Property Type for Module Metadata

### PR DESCRIPTION
**Pull request by gaurabdg**
#8801

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java [hunk 1]
```diff
@@ -170,30 +170,32 @@ private static void verifyInputFile(Map<String, List<String>> allTests, File fil
...
-            slash = packge.lastIndexOf(File.separatorChar);
-            packge = path.substring(0, slash);
-            classes = allTests.get(packge);
-
-            if (classes != null
-                    && checkInputMatchCorrectFileStructure(classes, folderPath, skipFileNaming,
-                            fileName)) {
-                found = true;
-                break;
+        if (!skipFileNaming) {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java:173]: please review PR one more time, and remove all unrelated changes.

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Comment by romani** [comment]
from Teamcity IDEA inspection:
```
src/main/java/com/puppycrawl/tools/checkstyle/meta
MetadataGeneratorUtil.java (1)
109: fetchAllTokens() Cast from 'double' to int may result in loss of precision 

```

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 2]
```diff
@@ -877,10 +878,13 @@ private static void validatePropertySectionPropertyTokens(String fileName, Strin
...
 
         final String acceptableTokenText = columns.get(2).getTextContent().trim();
-        assertEquals("subset of tokens "
-                        + CheckUtil.getTokenText(check.getAcceptableTokens(),
-                check.getRequiredTokens()),
-                acceptableTokenText
+        String expectedAcceptableTokenText = "subset of tokens "
+                + CheckUtil.getTokenText(check.getAcceptableTokens(),
+                check.getRequiredTokens());
+        if ("DescendantToken".equals(sectionName)) {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:884]:
```
tokens  = CheckUtil.getTokenText(check.getAcceptableTokens(),
                check.getRequiredTokens())
if (tokens.size() == TOKEN_TYPE_TOTAL_COUNT) {
 expectedAcceptableTokenText = "all tokens allTokenTypes";
} else {
 expectedAcceptableTokenText = "subset of tokens "
                + tokens;
}
```

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java [hunk 3]
```diff
@@ -745,7 +745,8 @@ else if (AbstractCheck.class.isAssignableFrom(clss)) {
             Arrays.sort(requiredTokens);
 
             if (!Arrays.equals(acceptableTokens, defaultTokens)
-                    || !Arrays.equals(acceptableTokens, requiredTokens)) {
+                    || !Arrays.equals(acceptableTokens, requiredTokens)
+                    || "DescendantToken".equals(sectionName)) {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:749]: ` || acceptableTokens.size() == TOKEN_TYPE_TOTAL_COUNT)`

but why we need this condition, `Arrays.equals` will never be true.
Please remove.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java [hunk 4]
```diff
@@ -317,32 +317,28 @@ private static String createPropertiesText() {
 
             String typeText = "java.lang.String[]";
             final String propertyType = property.get(2).getTextContent();
-            final boolean isPropertyTokenType = propertyType.contains("subset of tokens")
+            final boolean isSpecialAllTokensType = propertyType.contains("all tokens");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:320]: `set of any tokens` is more logical, sorry for confusion from the beginning.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java [hunk 5]
```diff
@@ -317,32 +317,28 @@ private static String createPropertiesText() {
...
-                else {
-                    typeNode = property.get(2).getFirstChild().getFirstChild();
-                }
-                final String typeName = typeNode.getTextContent().trim();
+                final String typeName =
+                        getCorrectNodeBasedOnPropertyType(property).getTextContent().trim();
                 typeText = FULLY_QUALIFIED_CLASS_NAMES.get(typeName).getTypeName();
             }
+            if (isSpecialAllTokensType) {
+                typeText = "allTokenTypes";
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java:330]: `anyTokenTypesSet`

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 2 above)
Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:884]: But `DescendantTokenCheck` will satisfy that `tokens.size() != TokenUtil.getAllTokenIds().length`, and the test will fail. tokens.size() is zero.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 3 above)
Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:749]: Why not? both acceptableTokens and defaultTokens is empty for DescendantTokenCheck, it is an exception for all the cases.

File: src/xdocs/config_misc.xml [hunk 6]
```diff
@@ -673,6 +673,16 @@ String unitAbbrev = "\u03bc\u03bc\u03bc";
               <td><code>null</code></td>
               <td>3.2</td>
             </tr>
+            <tr>
+              <td>tokens</td>
+              <td>tokens to check</td>
+              <td>all tokens
+                <a href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html">
+                    allTokenTypes</a>
```
[the remark this answers is not in the document]
Comment (gaurabdg) [reply src/xdocs/config_misc.xml]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java (hunk 4 above)
Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java (hunk 5 above)
Comment (gaurabdg) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsJavaDocsTest.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 3 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:749]: https://github.com/checkstyle/checkstyle/blob/3686c0f493e37831d1450b6ac608b9734fb7efa0/src/main/java/com/puppycrawl/tools/checkstyle/checks/DescendantTokenCheck.java#L477

acceptableTokens is not empty.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 2 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java:884]: `check.getRequiredTokens()` is empty for Check. getAcceptableTokens are all tokens. So `tokens` will be all tokens.

Am I miss something.

Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: fixed.
One logic miss hardcode ... new problem was found and fixed.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java (hunk 3 above)
Comment (romani) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java]: fixed.

**Review by romani — APPROVED** [review]
ok to merge ones CI pass
